toolchain/check/impl_lookup.h

```diff
@@ -35,6 +38,60 @@ auto LookupImplWitness(Context& context, SemIR::LocId loc_id,
                        SemIR::ConstantId query_facet_type_const_id)
     -> SemIR::InstBlockIdOrError;
 
+// The result of EvalLookupSingleImplWitness(). It can be one of:
+// - No value. Lookup failed to find an impl declaration.
+// - A concrete value. Lookup found a concrete impl declaration that can be
+//   used definitively.
+// - A symbolic value. Lookup found an impl but it is not returned since lookup
+//   will need to be done again with a more specific query to look for
+//   specializations.
+class EvalImplLookupResult {
+ public:
+  // NOLINTNEXTLINE(google-explicit-constructor)
+  EvalImplLookupResult(SemIR::InstId inst_id) : result_(inst_id) {}
+
+  static auto MakeNonFinalResult() -> EvalImplLookupResult {
+    return EvalImplLookupResult(FoundNonFinalImpl());
+  }
+
+  // True if a concrete impl witness was found or a non-final impl. In the
+  // latter case the InstId of the impl's witness is not returned, only the fact
+  // that it exists.
+  auto has_value() const -> bool {
+    return std::holds_alternative<FoundNonFinalImpl>(result_) ||
+           std::get<SemIR::InstId>(result_).has_value();
+  }
+
+  // True if there is a concrete witness in the result. If false, and
+  // `has_value()` is true, it means a non-final impl was found and a further
+  // more specific query will need to be done.
+  auto has_concrete_value() const -> bool {
+    return std::holds_alternative<SemIR::InstId>(result_);
+  }
+
+  // Only valid if `has_concrete_value()` is true. Returns the witness id for
+  // the found impl declaration, or None if `has_value()` is false.
+  auto GetConcreteWitness() const -> SemIR::InstId {
```

Comment:
Maybe rename to use accessor-style naming?

```suggestion
  auto concrete_witness() const -> SemIR::InstId {
```

Or consistent with `has_concrete_value`:

```suggestion
  auto concrete_value() const -> SemIR::InstId {
```

Reply:
Yeah I was torn here, since `GetConcreteWitness` felt like it should have a camel case, and `has_value` should have underscore case, there was no winning. If you feel like we should pick one, `has_value` has a stronger case so I will go with underscores.